java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java

```diff
@@ -0,0 +1,243 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.java.util.common.granularity;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonSerializable;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.jsontype.TypeSerializer;
+import io.druid.java.util.common.IAE;
+import io.druid.java.util.common.RE;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.Period;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import java.io.IOException;
+
+public class PeriodSegmentGranularity extends SegmentGranularity implements JsonSerializable
+{
+  private final GranularityType granularityType;
+  private final PeriodGranularity periodGranularity;
+
+  @JsonCreator
+  public PeriodSegmentGranularity(
+      @JsonProperty("period") Period period,
+      @JsonProperty("origin") DateTime origin,
+      @JsonProperty("timeZone") DateTimeZone tz
+  )
+  {
+    this.periodGranularity = new PeriodGranularity(period, origin, tz);
+    this.granularityType = GranularityType.estimatedGranularityType(period);
```

Comment:
Currently, segmentGranularity has to be one of the predefined values as per the Granularity enum i.e SECOND, MINUTE, FIVE_MINUTE, TEN_MINUTE etc 

With this PR, segmentGranularity can be represented with periods. But with one restriction, the periods can't be compound meaning 1hour5minutes etc. This is because compound periods makes it harder to match a "hive", "default", "lowerDefault" format representation. We don't have this format issue with custom periods that are not compound. For example: PT2S has the same format as PT1S. PT13M has the same format as PT1M.

So, that's why `GranularityType.estimatedGranularityType(period);` will throw `IAE` for compound periods. 

I have two tests for this: `testCustomPeriodToDate` and `testCustomNestedPeriodFail`.